swarm/runtime/statsdb/ingestion.py

```
@@ -551,10 +551,20 @@ def ingest_events(self, events: List[Dict[str, Any]], run_id: str) -> int:
-            return self._ingest_events_internal(events, run_id)
-        finally:
-            _ingestion_context.active = False
+
+        # Use a transaction for the entire batch to improve performance
+        # DuckDB (and SQLite) are much faster with a single transaction
+        with self._lock:
+            try:
+                self.connection.execute("BEGIN TRANSACTION")
+                result = self._ingest_events_internal(events, run_id)
+                self.connection.execute("COMMIT")
+                return result
```

Ruff (0.15.0)

[warning] Consider moving this statement to an `else` block

(TRY300)